homeassistant/components/loxone.py

```diff
@@ -0,0 +1,230 @@
+"""
+Component to create an interface to the Loxone Miniserver
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/loxone/
+"""
+# pylint: disable=unused-import, too-many-lines
+import asyncio
+import codecs
+import hashlib
+import hmac
+import logging
+import uuid
+import json
+from datetime import timedelta
+from struct import unpack
+
+import voluptuous as vol
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import (CONF_HOST, CONF_PORT, CONF_USERNAME,
+                                 CONF_PASSWORD)
+from homeassistant.helpers.event import async_track_time_interval
+
+REQUIREMENTS = ['websockets==3.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'loxone'
+DEFAULT_PORT = 80
+EVENT = 'loxone_received'
+
+KEEPALIVEINTERVAL = timedelta(seconds=120)
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+DEFAULT = ""
+ATTR_UUID = 'uuid'
+ATTR_VALUE = 'value'
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Set up the Loxone component."""
+    user = config[DOMAIN][CONF_USERNAME]
+    password = config[DOMAIN][CONF_PASSWORD]
+    host = config[DOMAIN][CONF_HOST]
+    port = config[DOMAIN][CONF_PORT]
+    api = LoxoneGateway(hass, user, password, host, port)
+    api.start_listener(api.get_process_message_callback())
+    async_track_time_interval(hass, api.send_keepalive, KEEPALIVEINTERVAL)
+
+    @asyncio.coroutine
+    def handle_event_bus_command(call):
+        """Handle event bus services."""
+        value = call.data.get(ATTR_VALUE, DEFAULT)
+        device_uuid = call.data.get(ATTR_UUID, DEFAULT)
+        data = {device_uuid: value}
+        hass.bus.async_fire(EVENT, data)
+
+    hass.services.async_register(DOMAIN, 'event_bus_command',
+                                 handle_event_bus_command)
+
+    @asyncio.coroutine
+    def handle_websocket_command(call):
+        """Handle websocket command services."""
+        value = call.data.get(ATTR_VALUE, DEFAULT)
+        device_uuid = call.data.get(ATTR_UUID, DEFAULT)
+        yield from api.send_websocket_command(device_uuid, value)
+
+    hass.services.async_register(DOMAIN, 'event_websocket_command',
+                                 handle_websocket_command)
+
+    return True
+
+
+class LoxoneGateway:
+    """ Main class for the communication with the miniserver """
```

Comment:
See [PEP257](https://www.python.org/dev/peps/pep-0257/) for details about proper docstrings and the docs about [testing](https://home-assistant.io/developers/development_testing/) them.